File: manifold_sampling/py/tests/TestCreatePiecewiseQuadraticHfun.py
"""
Unit test of create_piecewise_quadratic_hfun()

Claude Sonnet 4.5 via Argo was used to generate a starting point for this test.
It was provided with the current working versions of

1. create_piecewise_quadratic_hfun.m, and
2. create_censored_L1_loss_hfun.m,
3. TestCreateCensoredL1LossHfun.m

at that time and asked to build up this test of (1) so that it tests (1) in a
way analogous to how (3) tests (2).

That code was then reviewed, cleaned, and evolved with no further contributions
or alterations by a generative AI tool.  Developers are, as expected, taking
responsibility for the correctness of this content and its ability to test the
related code.
"""

import unittest

import numpy as np

from ibcdfo.manifold_sampling import create_piecewise_quadratic_hfun


class TestCreatePiecewiseQuadraticHfun(unittest.TestCase):
    def setUp(self):
        EPS = np.finfo(float).eps
        M, L = (3, 4)

        # Confirm that we have good arguments for use in testing
        #
        # Use Fortran ordering so that this creates the same problem as defined
        # in the MATLAB test suite.
        Qs = np.arange(1, M * M * L + 1).astype(float)
        self.__Qs = Qs.reshape((M, M, L), order="F")
        zs = np.arange(1, M * L + 1).astype(float)
        self.__zs = zs.reshape((M, L), order="F")
        self.__cs = np.arange(1, L + 1).astype(float)

        self.__hfun = create_piecewise_quadratic_hfun(self.__Qs, self.__zs, self.__cs)

        # The test values were not determined by hand and are, therefore, not
        # known to be correct.  Rather, they were gathered from test output at
        # some time and are used here to catch regressions and to confirm that
        # the Python and MATLAB code are returning the same values for the same
        # problems.
        self.__Z = 1.1 * np.arange(1, M + 1)
        self.__hF, self.__grads, self.__Hash = self.__hfun(self.__Z)
        self.assertEqual(self.__hF, 22285.6)
        expected = np.array([-1635.6, -1688.4, -1741.2])
        rel_diff = np.max(np.fabs(1.0 - np.squeeze(self.__grads) / expected))
        self.assertTrue(rel_diff <= 5.0 * EPS)
        self.assertEqual(len(self.__Hash), 1)
        self.assertEqual(self.__grads.shape, (M, 1))
        # In the MATLAB version of this test, the hash is "4".  This is due to
        # the fact that for this problem the hash result is the index to the
        # active quadratic piece and indices in MATLAB are 1-based instead of
        # 0-based.
        self.assertEqual(self.__Hash, ["3"])

        # Sanity check hash result/H0
        hF_H0, grads_H0 = self.__hfun(self.__Z, self.__Hash)
        self.assertEqual(hF_H0, self.__hF)
        self.assertTrue(np.array_equal(grads_H0, self.__grads))

        # Good but different size for testing size incompatibilities
        M_long = M + 1
        L_long = L - 1
        Qs = np.arange(1, M_long * M_long * L_long + 1).astype(float)
        self.__Qs_long = Qs.reshape((M_long, M_long, L_long), order="F")
        zs = np.arange(1, M_long * L_long + 1).astype(float)
        self.__zs_long = zs.reshape((M_long, L_long), order="F")
        self.__cs_long = np.arange(1, L_long + 1).astype(float)

        hfun = create_piecewise_quadratic_hfun(self.__Qs_long, self.__zs_long, self.__cs_long)

        Z_long = 2.1 * np.arange(1, M_long + 1)
        hF, grads, Hash = hfun(Z_long)
        self.assertEqual(hF, 17286.0)
        expected = np.array([-1594, -1636, -1678, -1720])
        rel_diff = np.max(np.fabs(1.0 - np.squeeze(grads) / expected))
        self.assertTrue(rel_diff <= 5.0 * EPS)
        self.assertEqual(len(Hash), 1)
        self.assertEqual(grads.shape, (M_long, 1))
        # Similar to above comment, the hash here is one less than the result
        # in the MATLAB version of this test.
        self.assertEqual(Hash, ["2"])

        # Sanity check hash result/H0
        hF_H0, grads_H0 = hfun(Z_long, Hash)
        self.assertEqual(hF_H0, hF)
        self.assertTrue(np.array_equal(grads_H0, grads))

    def testErrors(self):
        M, L = self.__zs.shape

        # Qs, zs & cs must be numpy arrays ...
        bad_all = [None, "bad", 1.1, [1.1], (1.1,), {1.1}]
        for bad in bad_all:
            with self.assertRaises(TypeError):
                create_piecewise_quadratic_hfun(bad, self.__zs, self.__cs)
            with self.assertRaises(TypeError):
                create_piecewise_quadratic_hfun(self.__Qs, bad, self.__cs)
            with self.assertRaises(TypeError):
                create_piecewise_quadratic_hfun(self.__Qs, self.__zs, bad)

        # Qs must be 3D ...
        bad_all = [
            np.array([]),
            np.array([[]]),
            np.array(1.1),
            np.array([1.1]),
            np.array([[1.1]]),
            self.__Qs[:, :, 0],
            np.stack((self.__Qs, self.__Qs), axis=3),
        ]
        for bad in bad_all:
            self.assertTrue(bad.ndim != 3)
            with self.assertRaises(ValueError):
                create_piecewise_quadratic_hfun(bad, self.__zs, self.__cs)

        # zs must be 2D
        bad_all = [
            np.array([]),
            np.array(1.1),
            np.array([1.1]),
            self.__zs[:, 0],
            np.stack((self.__zs, self.__zs), axis=2),
            np.atleast_3d(self.__zs.copy()),
        ]
        for bad in bad_all:
            self.assertTrue(bad.ndim != 2)
            with self.assertRaises(ValueError):
                create_piecewise_quadratic_hfun(self.__Qs, bad, self.__cs)

        # cs must be effectively 1D (i.e., allow for 2D row/column vectors)
        for bad in [np.array([]), np.stack((self.__cs, self.__cs), axis=1)]:
            with self.assertRaises(ValueError):
                create_piecewise_quadratic_hfun(self.__Qs, self.__zs, bad)

        cs_row = np.atleast_2d(self.__cs.copy()).T
        self.assertEqual(2, cs_row.ndim)
        self.assertNotEqual(1, cs_row.shape[0])
        hfun_r = create_piecewise_quadratic_hfun(self.__Qs, self.__zs, cs_row)
        hF_r, grads_r, Hash_r = hfun_r(self.__Z)
        self.assertEqual(hF_r, self.__hF)
        self.assertTrue(np.array_equal(grads_r, self.__grads))
        self.assertEqual(Hash_r, self.__Hash)

        cs_column = cs_row.T
        self.assertEqual(2, cs_column.ndim)
        self.assertNotEqual(1, cs_column.shape[1])
        hfun_c = create_piecewise_quadratic_hfun(self.__Qs, self.__zs, cs_column)
        hF_c, grads_c, Hash_c = hfun_c(self.__Z)
        self.assertEqual(hF_c, self.__hF)
        self.assertTrue(np.array_equal(grads_c, self.__grads))
        self.assertEqual(Hash_c, self.__Hash)

        # Finite reals please for Qs, zs, and cs
        for bad_value in [np.nan, np.inf, -np.inf, 1j, 1.0 - 2.0 * 1j]:
            for k in range(L):
                bad = self.__cs.copy()
                if isinstance(bad_value, complex):
                    bad = bad.astype(complex)
                bad[k] = bad_value
                with self.assertRaises(ValueError):
                    create_piecewise_quadratic_hfun(self.__Qs, self.__zs, bad)

                for i in range(M):
                    bad = self.__zs.copy()
                    if isinstance(bad_value, complex):
                        bad = bad.astype(complex)
                    bad[i, k] = bad_value
                    with self.assertRaises(ValueError):
                        create_piecewise_quadratic_hfun(self.__Qs, bad, self.__cs)

                    for j in range(M):
                        bad = self.__Qs.copy()
                        if isinstance(bad_value, complex):
                            bad = bad.astype(complex)
                        bad[i, j, k] = bad_value
                        with self.assertRaises(ValueError):
                            create_piecewise_quadratic_hfun(bad, self.__zs, self.__cs)

        # Qs, zs & cs must have compatible shapes
        # Qs must be square in first two dimensions
        bad_Qs = np.ones((M, M + 1, L))
        with self.assertRaises(ValueError):
            create_piecewise_quadratic_hfun(bad_Qs, self.__zs, self.__cs)

        # zs must match Qs dimensions
        bad_zs = np.ones((M + 1, L))
        with self.assertRaises(ValueError):
            create_piecewise_quadratic_hfun(self.__Qs, bad_zs, self.__cs)

        bad_zs = np.ones((M, L + 1))
        with self.assertRaises(ValueError):
            create_piecewise_quadratic_hfun(self.__Qs, bad_zs, self.__cs)

        # cs must match Qs third dimension
        bad_cs = np.ones(L + 1)
        with self.assertRaises(ValueError):
            create_piecewise_quadratic_hfun(self.__Qs, self.__zs, bad_cs)

        # Test incompatibility between different good arrays
        with self.assertRaises(ValueError):
            create_piecewise_quadratic_hfun(self.__Qs_long, self.__zs, self.__cs)
        with self.assertRaises(ValueError):
            create_piecewise_quadratic_hfun(self.__Qs, self.__zs_long, self.__cs)
        with self.assertRaises(ValueError):
            create_piecewise_quadratic_hfun(self.__Qs, self.__zs, self.__cs_long)

        # Confirm that l must be >= 2
        Qs_bad = np.atleast_3d(self.__Qs[:, :, 0])
        zs_bad = np.atleast_2d(self.__zs[:, 0]).T
        cs_bad = np.atleast_1d(self.__cs[0])
        # Confirm that they are mutually compatible so that error is just due to
        # l=1.
        self.assertTrue(Qs_bad.ndim == 3)
        self.assertEqual(Qs_bad.shape[0], Qs_bad.shape[1])
        self.assertEqual(1, Qs_bad.shape[2])
        self.assertTrue(zs_bad.ndim == 2)
        self.assertEqual(zs_bad.shape[0], Qs_bad.shape[0])
        self.assertEqual(1, zs_bad.shape[1])
        self.assertEqual(1, len(cs_bad))
        with self.assertRaises(ValueError):
            create_piecewise_quadratic_hfun(Qs_bad, zs_bad, cs_bad)

    def testBadZArguments(self):
        M = self.__zs.shape[0]

        # z must be numpy array
        bad_all = [None, "bad", 1.1, list(self.__Z), tuple(self.__Z), set(self.__Z)]
        for bad in bad_all:
            with self.assertRaises(AssertionError):
                self.__hfun(bad)

        # z must be 1D
        bad_all = [np.array([[]]), np.atleast_2d(self.__Z), np.stack((self.__Z, self.__Z), axis=1)]
        for bad in bad_all:
            with self.assertRaises(AssertionError):
                self.__hfun(bad)

        # z must have correct length
        for bad in [np.array([]), np.ones(M - 1), np.ones(M + 1)]:
            with self.assertRaises(ValueError):
                self.__hfun(bad)

        # z must contain finite real values
        for bad_value in [np.nan, np.inf, -np.inf, 1j, 1.0 - 2.0 * 1j]:
            bad_orig = np.random.randn(M)
            for i in range(M):
                bad = bad_orig.copy()
                if isinstance(bad_value, complex):
                    bad = bad.astype(complex)
                bad[i] = bad_value
                with self.assertRaises(ValueError):
                    self.__hfun(bad)

    def testConfirmReadonly(self):
        # NOTE: This test is only useful for developing and manually testing the
        # code since the function under test doesn't actually alter Qs, zs, or cs
        # and likely should never do so.
        #
        # Therefore, developers can, if so desired, temporarily inject changes
        # to Qs, zs, or cs in the function and run this test to confirm that the
        # function is written such that it is incapable of inadvertently altering
        # the Qs, zs, cs arrays created and managed here.
        Qs = self.__Qs.copy()
        zs = self.__zs.copy()
        cs = self.__cs.copy()
        hfun = create_piecewise_quadratic_hfun(Qs, zs, cs)
        hfun(self.__Z)
        self.assertTrue(np.array_equal(Qs, self.__Qs))
        self.assertTrue(np.array_equal(zs, self.__zs))
        self.assertTrue(np.array_equal(cs, self.__cs))

    def testConfirmImmutable(self):
        # Construct using variables declared in this scope & collect results
        Qs = self.__Qs.copy()
        zs = self.__zs.copy()
        cs = self.__cs.copy()
        hfun = create_piecewise_quadratic_hfun(Qs, zs, cs)
        hF, grads, Hash = hfun(self.__Z)

        # Alter same construction variables & confirm that they yield different
        # results
        Qs *= -2.3
        hfun_2 = create_piecewise_quadratic_hfun(Qs, zs, cs)
        hF_2, grads_2, Hash_2 = hfun_2(self.__Z)
        self.assertNotEqual(hF_2, hF)
        self.assertFalse(np.array_equal(grads_2, grads))
        self.assertFalse(np.array_equal(Hash_2, Hash))

        zs *= -1.1
        hfun_3 = create_piecewise_quadratic_hfun(Qs, zs, cs)
        hF_3, grads_3, Hash_3 = hfun_3(self.__Z)
        self.assertNotEqual(hF_3, hF)
        self.assertNotEqual(hF_3, hF_2)
        self.assertFalse(np.array_equal(grads_3, grads))
        self.assertFalse(np.array_equal(grads_3, grads_2))
        self.assertFalse(np.array_equal(Hash_3, Hash))
        # self.assertFalse(np.array_equal(Hash_3, Hash_2))

        cs *= -1.5
        hfun_4 = create_piecewise_quadratic_hfun(Qs, zs, cs)
        hF_4, grads_4, Hash_4 = hfun_4(self.__Z)
        self.assertNotEqual(hF_4, hF)
        self.assertNotEqual(hF_4, hF_2)
        self.assertNotEqual(hF_4, hF_3)
        self.assertFalse(np.array_equal(grads_4, grads))
        self.assertFalse(np.array_equal(grads_4, grads_2))
        # self.assertFalse(np.array_equal(grads_4, grads_3))
        self.assertFalse(np.array_equal(Hash_4, Hash))
        # self.assertFalse(np.array_equal(Hash_4, Hash_2))
        # self.assertFalse(np.array_equal(Hash_4, Hash_3))

        # Confirm that changing the construction variables didn't alter the
        # original function
        hF_new, grads_new, Hash_new = hfun(self.__Z)
        self.assertEqual(hF_new, hF)
        self.assertTrue(np.array_equal(grads_new, grads))
        self.assertTrue(np.array_equal(Hash_new, Hash))

File: manifold_sampling/m/general_nonsmooth_h_funs/TestCreatePiecewiseQuadraticHfun.m
% From MATLAB and the directory containing this file, execute
%     >> runtests
%
% If you would like to run this from a different folder that includes these
% tests as a subfolder, then from that folder execute
%     >> runtests("IncludeSubfolders", true)
% Other tests run during execution might require adding other folders to the
% path.
%
% To execute the test suite with coverage enabled and to generate an HTML-format
% coverage report, execute from /path/to/IBCDFO/pounders/m
%     >> runtests("IncludeSubfolders", true, "ReportCoverageFor", pwd)
%

% Claude Sonnet 4.5 via Argo was used to generate a starting point for this
% test.  It was provided with the current working versions of
% 1. create_piecewise_quadratic_hfun.m, and
% 2. create_censored_L1_loss_hfun.m,
% 3. TestCreateCensoredL1LossHfun.m
% at that time and asked to build up this test of (1) so that it tests (1) in a
% way analogous to how (3) tests (2).
%
% That code was then reviewed, cleaned, and evolved with no further
% contributions or alterations by a generative AI tool.  Developers are, as
% expected, taking responsibility for the correctness of this content and its
% ability to test the related code.

classdef TestCreatePiecewiseQuadraticHfun < matlab.unittest.TestCase
    properties
        Qs
        zs
        cs
        hfun
        Z
        hF
        grads
        Hash
        Qs_long
        zs_long
        cs_long
        path_orig
    end

    methods (TestMethodSetup)

        function setup(testCase)
            % tests need product_of_cells
            [here_path, ~, ~] = fileparts(mfilename('fullpath'));
            testCase.path_orig = addpath(fullfile(here_path, '..'));

            % Confirm that we have good arguments for use in testing
            M = 3;
            L = 4;

            % Create test data
            testCase.Qs = reshape(1:(M * M * L), [M, M, L]);
            testCase.zs = reshape(1:(M * L), [M, L]);
            testCase.cs = 1:L;
            testCase.cs = testCase.cs(:);

            [testCase.hfun] = create_piecewise_quadratic_hfun(testCase.Qs, ...
                                                              testCase.zs, ...
                                                              testCase.cs);

            % The test values were not determined by hand and are, therefore,
            % not known to be correct.  Rather, they were gathered from test
            % output at some time and are used here to catch regressions and to
            % confirm that the Python and MATLAB code are returning the same
            % values for the same problems.
            testCase.Z = 1.1 * (1:M);
            [testCase.hF, testCase.grads, testCase.Hash] = ...
                testCase.hfun(testCase.Z);
            expected = 22285.6;
            rel_diff = abs(1.0 - testCase.hF / expected);
            testCase.assertTrue(rel_diff <= 5 * eps);
            expected = [-1635.6; -1688.4; -1741.2];
            rel_diff = max(abs(1.0 - expected ./ testCase.grads));
            testCase.assertTrue(rel_diff <= 5 * eps);
            testCase.assertTrue(iscell(testCase.Hash));
            [tmp, n_hash] = size(testCase.Hash);
            testCase.assertEqual(tmp, 1);
            testCase.assertEqual(n_hash, 1);
            testCase.assertEqual(size(testCase.grads), [M, 1]);
            % In the Python version of this test, the hash is "3".  This is due
            % to the fact that for this problem the hash result is the index to
            % the active quadratic piece and indices in Python are 0-based
            % instead of 1-based.
            testCase.assertEqual(testCase.Hash, {'4'});

            % Sanity check hash result/H0
            [hF_H0, grads_H0] = testCase.hfun(testCase.Z, testCase.Hash);
            testCase.assertEqual(hF_H0, testCase.hF);
            testCase.assertEqual(grads_H0, testCase.grads);

            % Good but different size for testing size incompatibilities
            M_long = M + 1;
            L_long = L - 1;
            testCase.Qs_long = reshape(1:(M_long * M_long * L_long), ...
                                       [M_long, M_long, L_long]);
            testCase.zs_long = reshape(1:(M_long * L_long), ...
                                       [M_long, L_long]);
            testCase.cs_long = 1:(L_long);
            testCase.cs_long = testCase.cs_long(:);

            [hfun] = create_piecewise_quadratic_hfun(testCase.Qs_long, ...
                                                     testCase.zs_long, ...
                                                     testCase.cs_long);

            % As above in terms of correctness and use of expected results
            % here.
            Z_long = 2.1 * (1:length(testCase.zs_long));
            [hF, grads, Hash] = hfun(Z_long);
            testCase.assertEqual(hF, 17286.0);
            expected = [-1594; -1636; -1678; -1720];
            rel_diff = max(abs(1.0 - expected ./ grads));
            testCase.assertTrue(rel_diff <= 5 * eps);
            testCase.assertTrue(iscell(Hash));
            [tmp, n_hash] = size(Hash);
            testCase.assertEqual(tmp, 1);
            testCase.assertEqual(n_hash, 1);
            testCase.assertEqual(size(grads), [M_long, 1]);
            % Similar to above comment, the hash here is one more than the
            % result in the Python version of this test.
            testCase.assertEqual(Hash, {'3'});

            % Sanity check hash result/H0
            [hF_H0, grads_H0] = hfun(Z_long, Hash);
            testCase.assertEqual(hF_H0, hF);
            testCase.assertEqual(grads_H0, grads);
        end

    end

    methods (TestMethodTeardown)

        function restorePath(testCase)
            path(testCase.path_orig);
        end

    end

    methods (Test)

        function badArguments(testCase)
            M = size(testCase.zs, 1);
            L = size(testCase.zs, 2);

            % Qs must be a 3D array
            testCase.verifyError( ...
                @()create_piecewise_quadratic_hfun([], testCase.zs, testCase.cs), ...
                'MATLAB:validators:mustBeNonempty' ...
            );
            Qs_2D = ones([M, M]);
            testCase.verifyEqual(ndims(Qs_2D), 2);
            testCase.verifyError( ...
                @()create_piecewise_quadratic_hfun(Qs_2D, testCase.zs, testCase.cs), ...
                'POptUS:IncompatibleSizes' ...
            );
            Qs_4D = ones([M, M, L, 2]);
            testCase.verifyEqual(ndims(Qs_4D), 4);
            testCase.verifyError( ...
                @()create_piecewise_quadratic_hfun(Qs_4D, testCase.zs, testCase.cs), ...
                'MATLAB:validation:IncompatibleSize' ...
            );

            % zs must be a matrix (not 3D)
            testCase.verifyError( ...
                @()create_piecewise_quadratic_hfun(testCase.Qs, [], testCase.cs), ...
                'MATLAB:validators:mustBeNonempty' ...
            );
            zs_1D = ones([L 1]);
            testCase.verifyEqual(ndims(zs_1D), 2);
            testCase.verifyError( ...
                @()create_piecewise_quadratic_hfun(testCase.Qs, zs_1D, testCase.cs), ...
                'POptUS:IncompatibleSizes' ...
            );
            zs_1D = ones([1 L]);
            testCase.verifyEqual(ndims(zs_1D), 2);
            testCase.verifyError( ...
                @()create_piecewise_quadratic_hfun(testCase.Qs, zs_1D, testCase.cs), ...
                'POptUS:IncompatibleSizes' ...
            );
            zs_3D = ones([M M L]);
            testCase.verifyEqual(ndims(zs_3D), 3);
            testCase.verifyError( ...
                @()create_piecewise_quadratic_hfun(testCase.Qs, zs_3D, testCase.cs), ...
                'MATLAB:validation:IncompatibleSize' ...
            );

            % cs must be a (2D) vector
            testCase.verifyError( ...
                @()create_piecewise_quadratic_hfun(testCase.Qs, testCase.zs, []), ...
                'MATLAB:validators:mustBeNonempty' ...
            );
            testCase.verifyError( ...
                @()create_piecewise_quadratic_hfun(testCase.Qs, testCase.zs, 1.1), ...
                'POptUS:IncompatibleSizes' ...
            );
            cs_2D = ones([M, L]);
            testCase.verifyEqual(ndims(cs_2D), 2);
            testCase.verifyError( ...
                @()create_piecewise_quadratic_hfun(testCase.Qs, testCase.zs, cs_2D), ...
                'MATLAB:validation:IncompatibleSize' ...
            );

            % cs can be row or column vector
            cs_row = testCase.cs(:)';
            testCase.verifyTrue(iscolumn(testCase.cs));
            testCase.verifyTrue(isrow(cs_row));
            [hfun_c] = create_piecewise_quadratic_hfun(testCase.Qs, ...
                                                       testCase.zs, ...
                                                       cs_row);
            [hF_c, grads_c, Hash_c] = hfun_c(testCase.Z);
            testCase.assertEqual(hF_c, testCase.hF);
            testCase.assertEqual(grads_c, testCase.grads);
            testCase.assertEqual(Hash_c, testCase.Hash);

            % Finite real values only please
            for bad = [1j, 1.0 - 2.0 * 1j]
                for k = 1:L
                    cs_bad = testCase.cs;
                    cs_bad(k) = bad;
                    testCase.assertError( ...
                        @()create_piecewise_quadratic_hfun( ...
                            testCase.Qs, testCase.zs, cs_bad ...
                        ), ...
                        'MATLAB:validators:mustBeReal' ...
                    );

                    for i = 1:M
                        zs_bad = testCase.zs;
                        zs_bad(i, k) = bad;
                        testCase.assertError( ...
                            @()create_piecewise_quadratic_hfun( ...
                                testCase.Qs, zs_bad, testCase.cs ...
                            ), ...
                            'MATLAB:validators:mustBeReal' ...
                        );

                        for j = 1:M
                            Qs_bad = testCase.Qs;
                            Qs_bad(i, j, k) = bad;
                            testCase.assertError( ...
                                @()create_piecewise_quadratic_hfun( ...
                                    Qs_bad, testCase.zs, testCase.cs ...
                                ), ...
                                'MATLAB:validators:mustBeReal' ...
                            );
                        end
                    end
                end
            end

            for bad = [inf -inf nan]
                for k = 1:L
                    cs_bad = testCase.cs;
                    cs_bad(k) = bad;
                    testCase.assertError( ...
                        @()create_piecewise_quadratic_hfun( ...
                            testCase.Qs, testCase.zs, cs_bad ...
                        ), ...
                        'MATLAB:validators:mustBeFinite' ...
                    );

                    for i = 1:M
                        zs_bad = testCase.zs;
                        zs_bad(i, k) = bad;
                        testCase.assertError( ...
                            @()create_piecewise_quadratic_hfun( ...
                                testCase.Qs, zs_bad, testCase.cs ...
                            ), ...
                            'MATLAB:validators:mustBeFinite' ...
                        );

                        for j = 1:M
                            Qs_bad = testCase.Qs;
                            Qs_bad(i, j, k) = bad;
                            testCase.assertError( ...
                                @()create_piecewise_quadratic_hfun( ...
                                    Qs_bad, testCase.zs, testCase.cs ...
                                ), ...
                                'MATLAB:validators:mustBeFinite' ...
                            );
                        end
                    end
                end
            end

            % Qs, zs, and cs must have compatible sizes
            testCase.verifyError( ...
                @()create_piecewise_quadratic_hfun(testCase.Qs_long, testCase.zs, testCase.cs), ...
                'POptUS:IncompatibleSizes' ...
            );
            testCase.verifyError( ...
                @()create_piecewise_quadratic_hfun(testCase.Qs, testCase.zs_long, testCase.cs), ...
                'POptUS:IncompatibleSizes' ...
            );
            testCase.verifyError( ...
                @()create_piecewise_quadratic_hfun(testCase.Qs, testCase.zs, testCase.cs_long), ...
                'POptUS:IncompatibleSizes' ...
            );

            % Test mismatched dimensions more explicitly
            Qs_wrong = zeros(M + 1, M, L);
            testCase.verifyError( ...
                @()create_piecewise_quadratic_hfun(Qs_wrong, testCase.zs, testCase.cs), ...
                'POptUS:IncompatibleSizes' ...
            );

            Qs_wrong = zeros(M, M + 1, L);
            testCase.verifyError( ...
                @()create_piecewise_quadratic_hfun(Qs_wrong, testCase.zs, testCase.cs), ...
                'POptUS:IncompatibleSizes' ...
            );

            Qs_wrong = zeros(M, M, L + 1);
            testCase.verifyError( ...
                @()create_piecewise_quadratic_hfun(Qs_wrong, testCase.zs, testCase.cs), ...
                'POptUS:IncompatibleSizes' ...
            );

            zs_wrong = zeros(M + 1, L);
            testCase.verifyError( ...
                @()create_piecewise_quadratic_hfun(testCase.Qs, zs_wrong, testCase.cs), ...
                'POptUS:IncompatibleSizes' ...
            );

            zs_wrong = zeros(M, L + 1);
            testCase.verifyError( ...
                @()create_piecewise_quadratic_hfun(testCase.Qs, zs_wrong, testCase.cs), ...
                'POptUS:IncompatibleSizes' ...
            );

            cs_wrong = zeros(L + 1, 1);
            testCase.verifyError( ...
                @()create_piecewise_quadratic_hfun(testCase.Qs, testCase.zs, cs_wrong), ...
                'POptUS:IncompatibleSizes' ...
            );

            cs_wrong = zeros(1, L + 1);
            testCase.verifyError( ...
                @()create_piecewise_quadratic_hfun(testCase.Qs, testCase.zs, cs_wrong), ...
                'POptUS:IncompatibleSizes' ...
            );
        end

        function badZArgument(testCase)
            M = size(testCase.zs, 1);

            % Must be 1D
            for bad = ["bad", 1.1, 1j, []]
                testCase.verifyError( ...
                    @()testCase.hfun(bad), 'POptUS:IncompatibleSizes' ...
                );
            end

            Z_2D = zeros([M, 2]);
            testCase.verifyError( ...
                @()testCase.hfun(Z_2D), 'MATLAB:assertion:failed' ...
            );

            % z shape must match zs
            Z_short = zeros(M - 1, 1);
            Z_long = zeros(M + 1, 1);
            testCase.verifyError( ...
                @()testCase.hfun(Z_short), 'POptUS:IncompatibleSizes' ...
            );
            testCase.verifyError( ...
                @()testCase.hfun(Z_long), 'POptUS:IncompatibleSizes' ...
            );

            % z should only be finite real vector
            for bad = [1j, 1.0 - 2.0 * 1j]
                for i = 1:length(testCase.Z)
                    z_bad = testCase.Z;
                    z_bad(i) = bad;
                    testCase.verifyError( ...
                        @()testCase.hfun(z_bad), 'POptUS:NonrealValues' ...
                    );
                end
            end

            for bad = [inf -inf nan]
                for i = 1:length(testCase.Z)
                    z_bad = testCase.Z;
                    z_bad(i) = bad;
                    testCase.verifyError( ...
                        @()testCase.hfun(z_bad), 'POptUS:NonfiniteValues' ...
                    );
                end
            end
        end

        function confirmImmutable(testCase)
            % This test case is motivated by technical subtleties seen with
            % Python.

            % Construct using variable declared in this scope & collect results
            Qs = testCase.Qs;
            zs = testCase.zs;
            cs = testCase.cs;
            [hfun] = create_piecewise_quadratic_hfun(Qs, zs, cs);
            [hF, grads, Hash] = hfun(testCase.Z);

            % Alter same Qs variable and confirm different result
            Qs = -2.3 * Qs;
            [hfun_2] = create_piecewise_quadratic_hfun(Qs, zs, cs);
            [hF_2, grads_2, Hash_2] = hfun_2(testCase.Z);
            testCase.assertNotEqual(hF_2, hF);
            testCase.assertNotEqual(grads_2, grads);
            testCase.assertNotEqual(Hash_2, Hash);

            % Confirm that changing Qs didn't alter the original function
            [hF_new, grads_new, Hash_new] = hfun(testCase.Z);
            testCase.assertEqual(hF_new, hF);
            testCase.assertEqual(grads_new, grads);
            testCase.assertEqual(Hash_new, Hash);

            % Alter same zs variable and confirm different result
            zs(:, 1) = -1.1 * zs(:, 1);
            [hfun_3] = create_piecewise_quadratic_hfun(Qs, zs, cs);
            [hF_3, grads_3, Hash_3] = hfun_3(testCase.Z);
            testCase.assertNotEqual(hF_3, hF);
            testCase.assertNotEqual(hF_3, hF_2);
            testCase.assertNotEqual(grads_3, grads);
            testCase.assertNotEqual(grads_3, grads_2);
            testCase.assertNotEqual(Hash_3, Hash);
            testCase.assertNotEqual(Hash_3, Hash_2);

            % Confirm that changing zs didn't alter the original function
            [hF_new, grads_new, Hash_new] = hfun(testCase.Z);
            testCase.assertEqual(hF_new, hF);
            testCase.assertEqual(grads_new, grads);
            testCase.assertEqual(Hash_new, Hash);

            % Alter same cs variable and confirm different result
            cs(1) = 500.6 + cs(1);
            [hfun_4] = create_piecewise_quadratic_hfun(Qs, zs, cs);
            [hF_4, grads_4, Hash_4] = hfun_4(testCase.Z);
            testCase.assertNotEqual(hF_4, hF);
            testCase.assertNotEqual(hF_4, hF_3);
            testCase.assertNotEqual(grads_4, grads);
            testCase.assertNotEqual(grads_4, grads_3);
            testCase.assertNotEqual(Hash_4, Hash);
            testCase.assertNotEqual(Hash_4, Hash_3);

            % Confirm that changing cs didn't alter the original function
            [hF_new, grads_new, Hash_new] = hfun(testCase.Z);
            testCase.assertEqual(hF_new, hF);
            testCase.assertEqual(grads_new, grads);
            testCase.assertEqual(Hash_new, Hash);
        end

    end
end
